Add Agent Builder skills + dashboards/viz documentation (#5868)
elastic/docs-content: explore-analyze/ai-features/agent-builder/skills.md

Skills differ from the agent's base system prompt: the system prompt is always in context, while skills are loaded selectively. An agent can have access to many skills without loading all of them into the context window at once.

## How agents use skills

When an agent receives a message, it selects a skill if it determines that one of the available skills is relevant to the query based on the skill's name and description. If a skill activates, it provides:

- **Knowledge content**: domain-specific instructions written in Markdown that tell the agent how to approach the task.
- **Tools**: [Built-in tools](tools/builtin-tools-reference.md) or [custom tools](tools/custom-tools.md)  that the agent can call while the skill is active.

## Use cases

Use skills when you have domain-specific knowledge or procedures that multiple agents should follow consistently. Some examples:

- A {{product.security}} user asks about a suspicious host. The [`entity-analytics`](builtin-skills-reference.md#agent-builder-entity-analytics-skill) skill activates and guides the agent through finding the entity, analyzing its risk score, asset criticality, and behavioral history.

> - A {{product.observability}} user asks why a service is slow. The [`observability.rca`](builtin-skills-reference.md#agent-builder-observability-rca-skill) skill activates and runs structured root cause analysis across logs, traces, and metrics.

Reviewer: This should be replaced with the `investigate `skill.  The `rca` skill is marked as experimental and it performs a much more narrow investigation (logs only) than the `investigate` skill.  `rca` skill may be removed or merged into this skill later.

- **Associated tools**: The tools the skill should have access to (up to 100). Available under **Advanced options** in the UI. Corresponds to `tool_ids` in the API.

:::{image} images/create-new-skill.png
:alt: Create skill dialog showing fields for ID, Name, Description, Instructions, and Associated tools
:width: 550px
:screenshot:
:::

::::

::::{step} Save the skill

Select **Save**. The skill is added to your library and becomes available to assign to any agent.

::::

:::::

### Use the API

You can also create and manage skills programmatically using the [Skills API](kibana-api.md#skills-apis). The API supports all the same fields as the UI, plus `referenced_content`: additional named content blocks the agent can read selectively.

## Add skills to an agent

To add skills to the selected agent, go to **Customize > Skills** and select **Add skills**. You have two options:

- **Import from skill library**: Add a previously created skill from **Manage components > Skills** to this agent.
- **Create a skill**: Create a new skill and add it to this agent in one step. The new skill is also saved to **Manage components > Skills**.

> % TODO: Confirm whether all built-in skills are assignable to custom agents, or some are solution-scoped.

Reviewer 2: There are skills which are solution-dependant (so may not appear for all solutions on serverless, e.g. the `rca` or something skill is o11y...)

---
navigation_title: "Skills"
description: "Learn how Agent Builder skills extend agents with specialized knowledge and tools for specific task domains."
applies_to:
  stack: ga 9.4+
  serverless: ga
products:
  - id: elasticsearch
  - id: kibana
  - id: observability
  - id: security
  - id: cloud-serverless
---

# Skills in {{agent-builder}}

Skills are reusable instruction sets that give [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.

Tools are discrete operations the agent can invoke. Skills are higher-level capability packs that bundle tools, instructions, and context for a specific task domain. To learn more about tools, refer to [Tools](tools.md).

Skills differ from the agent's base system prompt: the system prompt is always in context, while skills are loaded selectively. An agent can have access to many skills without loading all of them into the context window at once.

## How agents use skills

> When an agent receives a query, it selects the most relevant skill based on the task. The skill activates and provides:

Reviewer 3: A skill is not **always** selected; the agent may decide not to load a skill. Does this detail need to be stated? 

```suggestion
When an agent receives a message, it selects a skill if it determines that one of the available skills is relevant to the query based on the skill's name and description. If a skill activates, it provides:
```

Author: it helps the user's mental model so 👍

- **Knowledge content**: domain-specific instructions written in Markdown that tell the agent how to approach the task.
- **Registry tools**: IDs of existing [built-in tools](tools/builtin-tools-reference.md) that the agent can call while the skill is active.

> - **Inline tools**: skill-bound tools that are only available while the skill is running and do not appear in the tools catalog.

Reviewer 3: `Inline tools` is an implementation detail that is only relevant to built-in skills. For user-created skills, they can only use builtin or custom tools from the registry. I would just mention that skills provide "Knowledge content" and "Tools".

```suggestion
- **Tools**: [Built-in tools](tools/builtin-tools-reference.md) or [custom tools](tools/custom-tools.md)  that the agent can call while the skill is active.
```

For the complete list, refer to [Built-in skills reference](builtin-skills-reference.md).

## Create a custom skill

Custom skills are saved to your skill library and can be assigned to any agent. You can create a skill from **Manage components > Skills**, or inline when adding skills to an agent from **Customize > Skills**.

:::{tip}
For guidance on writing effective descriptions and instructions, refer to [Skill creation guidelines](skill-creation-guidelines.md).
:::

:::::{stepper}
::::{step} Open the Create skill dialog

Go to **Manage components > Skills**, then select **Create a skill**.

::::

::::{step} Fill in the skill fields

Complete the fields in the **Create skill** dialog:

- **ID**: A unique identifier for the skill.
- **Name**: A human-readable name.
- **Description**: A short description of when the agent should use the skill. This field is always included in the agent's context, so it should be specific and semantically distinct from other skill descriptions.
- **Instructions**: The skill content in Markdown. Include trigger conditions, step-by-step instructions, examples, and edge cases. Corresponds to `content` in the API.

> - **Associated tools**: The built-in tools the skill should have access to (up to 100). Available under **Advanced options** in the UI. Corresponds to `tool_ids` in the API.

Reviewer 3: Not only built in, also custom tools.

```suggestion
- **Associated tools**: The tools the skill should have access to (up to 100). Available under **Advanced options** in the UI. Corresponds to `tool_ids` in the API.
```

---
navigation_title: "Skills"
description: "Learn how Agent Builder skills extend agents with specialized knowledge and tools for specific task domains."
applies_to:
  stack: ga 9.4+
  serverless: ga
products:
  - id: elasticsearch
  - id: kibana
  - id: observability
  - id: security
  - id: cloud-serverless
---

# Skills in {{agent-builder}}

> A skill is a reusable instruction set that gives [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Write your expertise once, then share it across any number of agents. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.

Author: ```suggestion
A skill is a reusable instruction set that gives [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise easy to maintain and share across your team.
```

> Skills enable you to write expertise once and share it across any number of agents. A skill is a reusable instruction set that gives [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.

Reviewer 4: ```suggestion
Skills are reusable instruction sets that give [agents](agent-builder-agents.md) specialized expertise for particular tasks. They enable you to define procedures for any number of agents.  Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.
```
Another option

Author: thanks I tweaked a little bit in https://github.com/elastic/docs-content/pull/5868/changes#r3106869141

> Skills enable you to write expertise once and share it across any number of agents. A skill is a reusable instruction set that gives [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.

Author: ```suggestion
A skill is a reusable instruction set that gives [agents](agent-builder-agents.md) specialized expertise for a particular type of task. Write your expertise once, then share it across any number of agents. Instead of embedding the same detailed instructions in every agent that needs them, you author a skill once and assign it wherever it's needed. This keeps agent configurations clean and makes expertise shareable across your team.
```

Tools are discrete operations the agent can invoke. Skills are higher-level capability packs that bundle tools, instructions, and context for a specific task domain. To learn more about tools, refer to [Tools](tools.md).

Skills differ from the agent's base system prompt: the system prompt is always in context, while skills are loaded selectively. An agent can have access to many skills without loading them all into the context window at once.

## How agents use skills

When an agent receives a message, it selects a skill if it determines that one of the available skills is relevant to the query based on the skill's name and description. If a skill activates, it provides:

- **Knowledge content**: domain-specific instructions written in Markdown that tell the agent how to approach the task.
- **Tools**: [Built-in tools](tools/builtin-tools-reference.md) or [custom tools](tools/custom-tools.md)  that the agent can call while the skill is active.

## Use cases

Use skills when you have domain-specific knowledge or procedures that multiple agents should follow consistently. Some examples:

> - A {{product.security}} user asks about a suspicious host. The [`entity-analytics`](builtin-skills-reference.md#agent-builder-entity-analytics-skill) skill activates and guides the agent through finding the entity, analyzing its risk score, asset criticality, and behavioral history.

Reviewer 4: ```suggestion
- An {{product.security}} user asks about a suspicious host. The [`entity-analytics`](builtin-skills-reference.md#agent-builder-entity-analytics-skill) skill activates and guides the agent through finding the entity, analyzing its risk score, asset criticality, and behavioral history.
```

> - A {{product.observability}} user asks why a service is slow or why an alert fired. The [`observability.investigation`](builtin-skills-reference.md#agent-builder-observability-investigation-skill) skill activates and diagnoses the issue across APM services and infrastructure.

Reviewer 4: ```suggestion
- An {{product.observability}} user asks why a service is slow or why an alert fired. The [`observability.investigation`](builtin-skills-reference.md#agent-builder-observability-investigation-skill) skill activates and diagnoses the issue across APM services and infrastructure.
```
